src/main/java/org/jenkinsci/plugins/docker/commons/credentials/DockerRegistryEndpoint.java

```diff
@@ -181,12 +190,44 @@ DockerRegistryToken getToken(Item context) {
             // shrug off this error and move on. We are matching with ID anyway.
         }
 
+        Authentication runAuth = context instanceof Queue.Task ?
+                Tasks.getAuthenticationOf((Queue.Task) context) : ACL.SYSTEM;
+        List<IdCredentials> candidates = new ArrayList();
+        candidates.addAll(CredentialsProvider.lookupCredentials(IdCredentials.class, context, runAuth, requirements));
+        if (runAuth != ACL.SYSTEM && context.getACL().hasPermission(runAuth, CredentialsProvider.USE_ITEM)) {
+            candidates.addAll(CredentialsProvider.lookupCredentials(IdCredentials.class, context, ACL.SYSTEM, requirements));
+        }
+
         // look for subtypes that know how to create a token, such as Google Container Registry
-        return AuthenticationTokens.convert(DockerRegistryToken.class, firstOrNull(CredentialsProvider.lookupCredentials(
-                IdCredentials.class, context, Jenkins.getAuthentication(),requirements),
+        return AuthenticationTokens.convert(DockerRegistryToken.class, firstOrNull(candidates,
                 allOf(AuthenticationTokens.matcher(DockerRegistryToken.class), withId(credentialsId))));
     }
 
+    /**
+     * Plugins that want to refer to a {@link IdCredentials} should do so via ID string,
+     * and use this method to resolve it and convert to {@link DockerRegistryToken}.
+     *
+     * @param context
+     *       If you are a build step trying to access DockerHub in the context of a build/job,
+     *       specify that build. Otherwise null. If you are scoped to something else, you might
+     *       have to interact with {@link CredentialsProvider} directly.
+     */
+    public @CheckForNull DockerRegistryToken getToken(Run context) {
```

Comment:
Why is this `public` if it is only called internally and from tests? Make it default access. If and when some other plugin actually needs to call this (apparently `docker-workflow` does not), a separate set of PRs can be made which demonstrates the usage.